parsers/php.c

```diff
@@ -1282,22 +1362,39 @@ static bool parseFunction (tokenInfo *const token, const tokenInfo *name)
 		skipOverParens (token);
 	}
 
-	/* PHP7 return type declaration or if parsing Zephir, skip function return
-	 * type hint */
+	/* PHP7 return type declaration or if parsing Zephir, gather function return
+	 * type hint to fill typeRef. */
 	if ((getInputLanguage () == Lang_php && token->type == TOKEN_COLON) ||
 	    (getInputLanguage () == Lang_zephir && token->type == TOKEN_OPERATOR))
 	{
-		do
+		if (arglist)
+			rtype = vStringNew ();
+
+		readToken (token);
+		if (rtype && token->type == TOKEN_QMARK)
+		{
+			vStringPut (rtype, '?');
```

Comment:
it probably doesn't change much as all functions should have parentheses and this `rtype` should always be created as `arglist` should always be non-null, but skipping the `?` token should probably be unconditional, and the conditional be moved to the `vStringPut()` line.

Reply:
Than you for reviewing.
I wonder I understand what you say correctly.

My understanding:

```
if (arglist)
   rtype = vString ();
```
The condtion can be moved to upper level like:
```
if (arglist && ((getInputLanguage () == Lang_php && token->type == TOKEN_COLON) ||
	    (getInputLanguage () == Lang_zephir && token->type == TOKEN_OPERATOR)))
{
```

So we can remove arglist and rtype related conditions in the block:
```
  rtype = vString (); 
  readToken (token);
  if (token->type TOKEN_QMARK)
  {
     vStrintPut (rtype, '?');
     readToken(token);
   }
...
```
Am I understading what you wrote correctly?

Reply:
No, my point would be that the parsing would be the same whether or not we're collecting rtype, so that in the event we'd be parsing a correct declaration and simply don't want to collect the value, the code would still do the same thing.

It currently is not very relevant as the only case I can think of where `rtype` is `NULL` is for invalid code; but my idea is that currently the function behaves the same parsing-wise regardless of whether we're collecting or not, and that seemed nice to me.

However, if we indeed wanted to cut off some parsing when not needed, we could indeed move the check to a higher level.  But that wasn't my point -- which doesn't mean it's not valid :wink:

Reply:
> No, my point would be that the parsing would be the same whether or not we're collecting rtype, so that in the event we'd be parsing a correct declaration and simply don't want to collect the value, the code would still do the same thing.

Oh, I see. I think I understand.
So the change will be:
```
 			rtype = vStringNew ();
 
 		readToken (token);
-		if (rtype && token->type == TOKEN_QMARK)
+		if (token->type == TOKEN_QMARK)
 		{
-			vStringPut (rtype, '?');
+			if (rtype)
+				vStringPut (rtype, '?');
 			readToken (token);
 		}
 		readQualifiedName (token, rtype, NULL);
```
I will push the change. I have just found one more crack about rtype. I will push for fixing it, too.

Reply:
Yes, that's what I meant :+1: